### PR TITLE
add superknob linkings and improve effects defaults

### PR DESCRIPTION
**Pull request by Be-ing**
fixing https://bugs.launchpad.net/mixxx/+bug/1466966

### CODE REVIEW

**Comment by Be-ing** [comment]
@timrae : Want to test this?

**Comment by timrae** [comment]
Nice, let me give it a whirl

File: src/effects/native/flangereffect.cpp
Comment (timrae) [inline src/effects/native/flangereffect.cpp]: I think you should set the default delay much higher... Between 1000 and 5000

File: src/effects/native/echoeffect.cpp
Comment (timrae) [inline src/effects/native/echoeffect.cpp]: I think it makes more sense to set the feedback as the super parameter (without inversion) and put the default delay to something more sensible like 1s. Ideally you should have the delay synced to the bpm by default.

**Review by timrae — CHANGES_REQUESTED** [review]
Thanks for this, I made some comments inline!

File: src/effects/effectparameter.h
Comment (Be-ing) [inline src/effects/effectparameter.h]: EffectParameter only inherits from QObject.

File: src/effects/native/flangereffect.cpp
Comment (Be-ing) [inline src/effects/native/flangereffect.cpp]: done

File: src/effects/native/reverbeffect.cpp
Comment (Be-ing) [inline src/effects/native/reverbeffect.cpp]: I disagree on this one. I think the superknob should generally cause a more audible effect on the right and least audible/no effect on the left (except for bipolar superknobs like for filter).

File: src/effects/effectparameter.h
Comment (timrae) [inline src/effects/effectparameter.h]: Ah right sorry I was skim reading

File: src/effects/effectmanifestparameter.h
Comment (timrae) [inline src/effects/effectmanifestparameter.h]: Well you don't need to go through the whole file to fix the new ones... Plain enums are just [bad](http://stackoverflow.com/questions/18335861/why-is-enum-class-preferred-over-plain-enum).

File: src/effects/effectparameter.h
Comment (timrae) [inline src/effects/effectparameter.h]: I have no idea why all of the get/set functions in here are `virtual`... Why would someone want to override a get/set method? That's why I thought you were overriding it here, because otherwise it doesn't make sense to use the `virtual` keyword. I know you're just blending in with the rest of the code in the class, but just because the rest of the code is a bit sloppy, it doesn't mean that you should follow suit IMO.

File: src/effects/native/flangereffect.cpp
Comment (timrae) [inline src/effects/native/flangereffect.cpp]: Ugh, sorry I meant the period not the delay.

File: src/effects/native/reverbeffect.cpp
Comment (timrae) [inline src/effects/native/reverbeffect.cpp]: Isn't the bandwidth here just referring to a low-pass filter in the input? In this case a low-bandwidth means more filtering and a higher bandwidth means no filtering... According to my ears, the effect (at least of the filter) is much more audible with the knob at zero than it is at max.

Comment (Be-ing) [inline src/effects/native/reverbeffect.cpp]: I played around with it a bit more, and yeah, you're right.

IMO we could use some better sounding reverb effects, if anyone is up to the challenge.

File: src/effects/effectparameter.h
Comment (Be-ing) [inline src/effects/effectparameter.h]: From the comment at the top of effectmanifest.h:

> All methods are left virtual to allow a backend to replace the entire functionality with its own (for example, a database-backed manifest)

File: src/effects/native/flangereffect.cpp
Comment (Be-ing) [inline src/effects/native/flangereffect.cpp]: done

File: src/effects/effectparameter.h
Comment (timrae) [inline src/effects/effectparameter.h]: OK, thanks for the quote. I'm not really convinced that's the best design, but it does make sense to follow the existing code base if there's a valid reason to do so.

**Comment by timrae** [comment]
Thank you, I have reviewed and tested the code and I think it's ready to merge, though I would personally squash it down to one commit as 13 is a bit excessive ;)

**Comment by Be-ing** [comment]
I think this is ready to merge as well, unless anyone else has feedback.

We don't typically squash down PRs for Mixxx, but I could if requested.

**Comment by rryan** [comment]
Code LGTM, thanks for the change! (I haven't played with them)

GH supports squashing when you hit the merge button, but I tend to prefer preserving history. It helps for code archaeology (if I need to figure out why someone did something it can be helpful to look at how the hunk developed on their feature branch).

**Comment by timrae** [comment]
> if I need to figure out why someone did something it can be helpful to look at how the hunk developed on their feature branch.

Well if the code is reviewed properly then any unclear points should be clarified with comments ;) personally I believe if you have an unwieldy number of commits it makes harder to see why someone did something because you have to sift through a lengthy commit  history.

More importantly it makes it frustrating for contributors when they go through the commit history to see what's been happening, because you have to read and think about each commit message to see if it's something new or not.

**Comment by daschuer** [comment]
Nice improvement. I have just tested is and it works much better than master. 
I only do not like the Echo effect behavior. With the new default, the echo stacks up until clipping. 
Maybe we find a default setting that does not clip. Or is clipping a desired sound? 

It also sound like the effect clips the simples itselves. This disables unfortunately the clipping warning light, so the user has no visual clipping warning. 
Since Mixxx is a floating point engine, can't we just remove the clamping code inside the effect?

**Comment by Be-ing** [comment]
>  I only do not like the Echo effect behavior. With the new default, the echo stacks up until clipping.

I didn't notice until you point this out. Yes, it does eventually clip if you leave the metaknob/superknob all the way to the right for a while. But it sounds awful before that, so I always turned the knob left long before it got to clipping. I think most reasonable DJs would do the same. If you have better suggestions, I'm open to trying them.

**Comment by daschuer** [comment]
It could be worth to try to remove the clamping code from the echo effect. This will hopefully lit the clipping indicator as a visual warning.

**Comment by Be-ing** [comment]
Sure, but I think that's beyond the scope of this PR.

**Comment by Be-ing** [comment]
Ready for merge?

**Comment by daschuer** [comment]
> It could be worth to try to remove the clamping code from the echo effect. This will hopefully lit the clipping indicator as a visual warning.

Thinking of it again, I can remember that @ywwg wants to keep the clamping, to use the clipping sound as part of the effect.

**Comment by daschuer** [comment]
Thank you for fixing this long open bug! LGTM

**Comment by Be-ing** [comment]
Thanks for reviewing and testing!

**Pull request merged**
